### PR TITLE
Resizing issues

### PR DESCRIPTION
**Pull request by vampcat**
- Fixes the resizing issues.
- Replaced all explicit bind calls with StateChanges
- Some code refactoring to make the code more consistent across files

### How to test
Run the game, try resizing the game in windowed mode or going from fullscreen to windowed or vice versa (I've noticed an issue with the game sometimes not being able to go from windowed to fullscreen, but that's an issue outside the scope of this PR). Everything else related to rendering should be same (no weird water bugs, no screenshot crashes, no new visual artifacts).

This PR fixes the following:
#2744 
#2659
#2591
#1747
And quite possibly a few others related to resizing.

### CODE REVIEW

**Comment by vampcat** [comment]
Pinging @emanuele3d and @Cervator for review!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Awesome! This is way improved over the buggy previous state. I *did* still find ways to provoke the quirks, but it seems way harder, and may only happen after reloading an existing world?

I don't have exact reproducing steps but I would start the game then resize a bunch, toggle between screen modes (fullscreen etc), exit to main menu, load the game back up, then resize some more. One time after a reload the game looked fine but resizing would reproduce the old blackness in the resized areas. Two times *immediately* after reloading a world it came out immediately off ratio or something, like so:

2017-03-23 00_09_07-terasology - c__dev_terasology_git_terasology - intellij idea 2016 3 3

**Comment by vampcat** [comment]
Okay, that's definitely not good. I'll be travelling through that part of the code for quite some time, so I'll keep an out for anything else that might cause these issues :D

File: engine/src/main/java/org/terasology/rendering/dag/nodes/FirstPersonViewNode.java [hunk 1]
```diff
@@ -69,10 +71,13 @@ public void initialise() {
...
 
         addDesiredStateChange(new SetViewportToSizeOf(READONLY_GBUFFER, displayResolutionDependentFBOs));
-        sceneOpaqueFbo = displayResolutionDependentFBOs.get(READONLY_GBUFFER);
+        addDesiredStateChange(new BindFBO(READONLY_GBUFFER, displayResolutionDependentFBOs));
+        update();
 
         // this guarantee the objects drawn by this node are always drawn in front of everything else
         addDesiredStateChange(new SetDepthFunction(GL_ALWAYS));
+
+        displayResolutionDependentFBOs.subscribe(this);
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/FirstPersonViewNode.java]: Let's move this up please.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/FirstPersonViewNode.java [hunk 2]
```diff
@@ -114,4 +117,9 @@ public void process() {
 
             PerformanceMonitor.endActivity();
     }
+
+    @Override
+    public void update() {
+        readOnlyGBufferFbo = displayResolutionDependentFBOs.get(READONLY_GBUFFER);
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/FirstPersonViewNode.java]: I might be wrong but this looks like another case where we are not actually using this and readOnlyGBufferFbo is not necessary?

File: engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueBlocksNode.java [hunk 3]
```diff
@@ -83,10 +81,13 @@ public void initialise() {
...
 
         addDesiredStateChange(new SetViewportToSizeOf(READONLY_GBUFFER, displayResolutionDependentFBOs));
-        sceneOpaqueFbo = displayResolutionDependentFBOs.get(READONLY_GBUFFER);
+        addDesiredStateChange(new BindFBO(READONLY_GBUFFER, displayResolutionDependentFBOs));
+        update();
 
         addDesiredStateChange(new EnableMaterial(CHUNK_SHADER.toString()));
         chunkShader = getMaterial(CHUNK_SHADER);
+
+        displayResolutionDependentFBOs.subscribe(this);
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueBlocksNode.java]: Let's move this up please.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueBlocksNode.java [hunk 4]
```diff
@@ -68,7 +66,7 @@
     private Material chunkShader;
     private SetWireframe wireframeStateChange;
     private RenderingDebugConfig renderingDebugConfig;
-    private FBO sceneOpaqueFbo;
+    private FBO readOnlyGBufferFbo;
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueBlocksNode.java]: It appears this variable might be unnecessary in this node.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueObjectsNode.java [hunk 5]
```diff
@@ -57,7 +59,7 @@
     private Camera playerCamera;
     private SetWireframe wireframeStateChange;
     private RenderingDebugConfig renderingDebugConfig;
-    private FBO sceneOpaqueFbo;
+    private FBO readOnlyGBufferFbo;
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueObjectsNode.java]: Same as above: it doesn't look like we are using this.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueObjectsNode.java [hunk 6]
```diff
@@ -72,7 +74,10 @@ public void initialise() {
         addDesiredStateChange(new LookThrough(playerCamera));
 
         addDesiredStateChange(new SetViewportToSizeOf(READONLY_GBUFFER, displayResolutionDependentFBOs));
-        sceneOpaqueFbo = displayResolutionDependentFBOs.get(READONLY_GBUFFER);
+        addDesiredStateChange(new BindFBO(READONLY_GBUFFER, displayResolutionDependentFBOs));
+        update();
+
+        displayResolutionDependentFBOs.subscribe(this);
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/OpaqueObjectsNode.java]: Let's move this up please.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/OverlaysNode.java [hunk 7]
```diff
@@ -59,7 +57,7 @@
 
     private Camera playerCamera;
     private SetWireframe wireframeStateChange;
-    private FBO sceneOpaqueFbo;
+    private FBO readOnlyGBuffer;
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/OverlaysNode.java]: Same as above: variable seemingly not in use.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java [hunk 8]
```diff
@@ -81,10 +81,9 @@ public void initialise() {
         playerCamera = worldRenderer.getActiveCamera();
         addDesiredStateChange(new LookThrough(playerCamera));
 
-        requiresFBO(new FBOConfig(REFRACTIVE_REFLECTIVE, FULL_SCALE, FBO.Type.HDR).useNormalBuffer(), displayResolutionDependentFBOs);
+        refractiveReflectiveFbo = requiresFBO(new FBOConfig(REFRACTIVE_REFLECTIVE, FULL_SCALE, FBO.Type.HDR).useNormalBuffer(), displayResolutionDependentFBOs);
         addDesiredStateChange(new BindFBO(REFRACTIVE_REFLECTIVE, displayResolutionDependentFBOs));
-        sceneOpaqueFbo = displayResolutionDependentFBOs.get(READONLY_GBUFFER);
-        refractiveReflectiveFbo = displayResolutionDependentFBOs.get(REFRACTIVE_REFLECTIVE);
+        readOnlyGBufferFbo = displayResolutionDependentFBOs.get(READONLY_GBUFFER);
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java]: Do we need this line here? readOnlyGBufferFbo seems to be used only in update(), where it is refetched.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java [hunk 9]
```diff
@@ -81,10 +81,9 @@ public void initialise() {
...
 
         addDesiredStateChange(new EnableMaterial(CHUNK_SHADER.toString()));
         chunkShader = getMaterial(CHUNK_SHADER);
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java]: Let's move the subscribe line below up, where the FBOs are.

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java [hunk 10]
```diff
@@ -114,20 +112,13 @@ public StateChange getDefaultInstance() {
...
             switch (textureType) {
-
                 case ColorTexture:
                     return inputFbo.colorBufferTextureId;
 
-                case DepthRenderBuffer:
-                    return inputFbo.depthStencilRboId;
-
-                case DepthTexture:
+                case DepthStencilTexture:
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: Why the removal of a case here?

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java [hunk 11]
```diff
@@ -136,36 +127,29 @@ private int fetchTextureId() {
...
-
-        inputFbo = fbosManager.get(fboURN);
-
-        // If a node taking advantage of this state change is disabled when a game is started, task is null.
-        // This is due to the task list generator not having yet called the generateTask() method.
-        if (task != null) {
-            task.setTextureId(fetchTextureId());
-        }
+        inputFbo = fboManager.get(fboUrn);
+        task.setTextureId(fetchTextureId());
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: Why the removal of the comment and the if block?

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java [hunk 12]
```diff
@@ -136,36 +127,29 @@ private int fetchTextureId() {
...
     }
 
     @Override
     public String toString() {
         if (this != defaultInstance) {
             return String.format("%30s: slot %s, fbo %s, textureType %s, fboManager %s, material %s, parameter '%s'", this.getClass().getSimpleName(),
-                    textureSlot, fboURN.toString(), textureType.name(), fbosManager.toString(), materialURN.toString(), parameterName);
+                    textureSlot, fboUrn.toString(), textureType.name(), fboManager.toString(), materialURN.toString(), parameterName);
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: fbosManager -> fboManager: picky, picky, picky. 

:smile:

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java [hunk 13]
```diff
@@ -193,27 +177,24 @@ public boolean equals(Object other) {
...
-        private String parameterName;
 
         /**
          * Constructs an instance of this inner class, according to the given parameters.
          *
          * @param textureSlot an integer indirectly identifying a texture unit on the GPU (textureUnit = GL_TEXTURE0 + textureSlot).
          * @param textureId the opengl id of a texture, usually obtained via glGenTextures().
          * @param materialURN an URN identifying a material.
-         * @param parameterName the name of a variable in the shader program used to sample the texture.
+         * @param materialParameter a string
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: Why the change of name and the shortening of the parameter description?

File: engine/src/main/java/org/terasology/rendering/opengl/AbstractFBOsManager.java [hunk 14]
```diff
@@ -259,7 +262,13 @@ public FBOConfig getFboConfig(ResourceUrn fboName) {
      */
     @Override
     public boolean subscribe(FBOManagerSubscriber subscriber) {
-        return fboManagerSubscribers.add(subscriber);
+        // TODO: Add sanity checks to ensure no object can resubscribe
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/opengl/AbstractFBOsManager.java]: I'd like this to happen in this PR please.

File: engine/src/main/java/org/terasology/rendering/opengl/AbstractFBOsManager.java [hunk 15]
```diff
@@ -259,7 +262,13 @@ public FBOConfig getFboConfig(ResourceUrn fboName) {
...
+        // return fboManagerSubscribers.add(subscriber);
+
+        // VAMPCAT : TODO: RESTORE!!!
+        if (!fboManagerSubscribers.add(subscriber))
+            return false;
+        return true;
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/opengl/AbstractFBOsManager.java]: Let's use a proper if/else statement here please.

File: engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java [hunk 16]
```diff
@@ -15,6 +15,7 @@
  */
 package org.terasology.rendering.world;
 
+import org.terasology.rendering.dag.*;
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java]: No star imports please.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/ApplyDeferredLightingNode.java [hunk 17]
```diff
@@ -57,22 +54,37 @@
...
         addDesiredStateChange(new SetInputTexture(
-                textureSlot++, sceneOpaqueFbo.depthStencilTextureId,  DEFERRED_LIGHTING_MATERIAL, "texSceneOpaqueDepth"));
+                textureSlot++, readOnlyGBufferFbo.depthStencilTextureId,  DEFERRED_LIGHTING_MATERIAL, "texSceneOpaqueDepth"));
         addDesiredStateChange(new SetInputTexture(
-                textureSlot++, sceneOpaqueFbo.normalsBufferTextureId, DEFERRED_LIGHTING_MATERIAL, "texSceneOpaqueNormals"));
+                textureSlot++, readOnlyGBufferFbo.normalsBufferTextureId, DEFERRED_LIGHTING_MATERIAL, "texSceneOpaqueNormals"));
         addDesiredStateChange(new SetInputTexture(
-                textureSlot,   sceneOpaqueFbo.lightBufferTextureId,   DEFERRED_LIGHTING_MATERIAL, "texSceneOpaqueLightBuffer"));
+                textureSlot,   readOnlyGBufferFbo.lightBufferTextureId,   DEFERRED_LIGHTING_MATERIAL, "texSceneOpaqueLightBuffer"));
+*/
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/ApplyDeferredLightingNode.java]: Ah, it was removed in the next PR. I'll take care of it here too.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/ApplyDeferredLightingNode.java [hunk 18]
```diff
@@ -85,8 +97,7 @@ public void initialise() {
     public void process() {
         PerformanceMonitor.startActivity("rendering/applyDeferredLighting");
 
-        sceneOpaquePingPongFbo.bind(); // TODO: remove and replace with a state change
-        sceneOpaquePingPongFbo.setRenderBufferMask(true, true, true);
+        writeOnlyGBufferFbo.setRenderBufferMask(true, true, true);
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/ApplyDeferredLightingNode.java]: That won't be needed, I converted this to a StateChange in the next PR.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropNode.java [hunk 19]
```diff
@@ -86,10 +81,9 @@ public void initialise() {
         new WireframeTrigger(renderingDebugConfig, this);
 
         addDesiredStateChange(new SetViewportToSizeOf(WRITEONLY_GBUFFER, displayResolutionDependentFBOs));
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropNode.java]: I planned on doing this either way after asking for your opinion, glad you mentioned it on your own! :D

File: engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropNode.java [hunk 20]
```diff
@@ -123,12 +119,11 @@ public void disableWireframe() {
...
         PerformanceMonitor.startActivity("rendering/backdrop");
 
-        sceneOpaqueFbo.bind(); // TODO: remove when we can bind this via a StateChange
-        sceneOpaqueFbo.setRenderBufferMask(true, false, false);
+        readOnlyGBufferFBO.setRenderBufferMask(true, false, false);
 
         glCallList(skySphere); // Draws the skysphere
 
-        sceneOpaqueFbo.setRenderBufferMask(true, true, true); // TODO: handle these via new StateChange to be created
+        readOnlyGBufferFBO.setRenderBufferMask(true, true, true); // TODO: handle these via new StateChange to be created
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropNode.java]: Adding to what I wrote in an earlier comment: The next PR takes care of this :D

File: engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropReflectionNode.java [hunk 21]
```diff
@@ -50,7 +50,6 @@
  *
  */
 public class BackdropReflectionNode extends AbstractNode {
-
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropReflectionNode.java]: I usually do it manually.. Should I be avoiding that? I do it mostly to maintain consistency between files.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/DeferredPointLightsNode.java [hunk 22]
```diff
@@ -92,7 +85,12 @@ public void initialise() {
 
         addDesiredStateChange(new DisableDepthTest());
 
+        addDesiredStateChange(new BindFBO(READONLY_GBUFFER, displayResolutionDependentFBOs));
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/DeferredPointLightsNode.java]: Not really. I'll add it for now for the sake of consistency, but since the default viewPort is the size of screen, I would recommend that we remove all setViewportToSizeOf(READONLY_GBUFFER) StateChanges, seeing how redundant they are.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java [hunk 23]
```diff
@@ -58,28 +57,24 @@ public void initialise() {
...
-        this.outputFBOconfig = outputConfig;
-        this.outputFBOmanager = outputManager;
-
-        requiresFBO(inputConfig, inputManager);
-        requiresFBO(outputConfig, outputManager);
-
-        outputFBO = outputManager.get(this.outputFBOconfig.getName());
-        outputManager.subscribe(this);
+        requiresFBO(inputFboConfig, inputFboManager);
+        outputFbo = requiresFBO(outputFboConfig, outputFboManager);
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java]: I was the one who added the update() function to other nodes, most of them didn't even have an update() earlier.
However, the reason I have not used update() in CopyImageToHMDNode and in DownSamplerNode is because it isn't applicable.

Over here, requiresFbo() creates the Fbo, and returns it's FboId. It would be inefficient to ignore this return value, and then ask the fboManager for it's fboId - hence i broke the pattern at times.
Similarly in CopyImageToHMDNode, leftEye and rightEye are generated using requireFbo(), no need to query them again using update().

File: engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java [hunk 24]
```diff
@@ -113,9 +110,6 @@ public void process() {
 
     @Override
     public void update() {
-        // Note: we don't need to update the inputFBO because only SetInputTextureFromFBO
-        // needs to keep it up to date and takes care of that internally.
-        outputFBO = outputFBOmanager.get(outputFBOconfig.getName());
+        outputFbo = outputFboManager.get(outputFboUrn);
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java]: That comment would have made sense, had there been an inputFbo variable.

(I removed it, because it was useless)

File: engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java (hunk 8 above)
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java]: No, this line is essential here. The reason is, readOnlyGBufferFbo has not been initialized till now. It can either be manually initialized here (as I've done) or it can be initialized by calling update() here (not recommended, since that re-fetches refractiveReflectiveFbo again).

Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java]: Although since the cost of re-fetching the fboId isn't too high in the grand scheme of things, I guess we can call update() in such nodes to maintain a pattern.

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 10 above)
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: The enum (and the corresponding retrieving function) are used for a BindTexture StateChange, so it should only support Textures. depthRenderBuffer is NOT a texture, and so logically shouldn't even be in the list. Even if I return the depthRenderBuffer Fbo, you can't just use it like the other returned values (in a glBindTexture(TEXTURE_2D, ...) call). So, for the sake of clear and logical code, I removed it.

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 11 above)
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: Because, that was just an edge case that had to be handled because of badly designed code. As things stand, the update() function will only be called after subscription to fboManager, and the subscription fo fboManager takes place only after creation of task (in generateDefaultTask). As things stand, task can never be null inside update(), no matter what the conditions.

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 12 above)
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: I hope you don't mind too much? I like clear code, and clear and informative names.

In this case, "Singular noun + Manager" should be the pattern (It's never sheepsManager, it's always sheepManager),and so I did it. I plan on renaming the FBOsManagers to FBOManagers in a separate PR too.

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 13 above)
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: Okay, I can't figure out myself what I was thinking. I'm sorry, I'll revert it back.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropReflectionNode.java (hunk 21 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/BackdropReflectionNode.java]: If you did it manually this in particular is a matter of personal preferences and I'm not too bothered either ways. The important thing is that it's not some kind of automated functionality you are unaware of. So, no worries on this one.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/DeferredPointLightsNode.java (hunk 22 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/DeferredPointLightsNode.java]: That's a good point: agreed. Let's remove them completely please.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java (hunk 23 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java]: Sorry, I didn't explain myself clearly enough. I'm thinking the initialization pattern:

`fbo = requiresFBO(config, manager);`

instead of using update(), could be more widespread. For example, have a look in DeferredPointLightsNode. I think it could be applicable there, to initialize the readOnlyGBuffer variable. And the same could be done in other nodes, skipping the update() call within initialize.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java (hunk 24 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java]: Sorry, I got confused here. You are right, it is no longer relevant.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java (hunk 8 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java]: I'm not sure I'm understanding your first comment. That variable is only needed/used in update() from what I can see. The FBO is instantiated both CPU/GPU side in requiresFBO if I remember correctly. =?

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 10 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: Understood, so it was actually a bug to keep it there as it wouldn't have worked with glBindTexture.

Can you put a comment about it? Something like:

```java
// depthStencilRboId is a possible FBO attachment but is not covered by a case here 
// as it wouldn't work with the glBindTexture(TEXTURE_2D) call.
```

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 12 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: No, I don't mind, if I did I would have been explicit about changing it. :wink:

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 11 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: I see what you mean. Good catch!

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 13 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: I think it's inaccurate. The parameter is really on the shader, not on the material. And it's not really the parameter itself, but its name. I guess the most accurate name would be shaderParameterName. But I thought parameterName + the description was enough.

**Comment by emanuele3d** [comment]
I responded to a number of "outdated" comments. Some require action, some not.

I did have a look at the new commits and I saw that they addressed most issues, thank you.

I think we are about 1-2 rounds of communication away from merging. Good job so far!

File: engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java (hunk 23 above)
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java]: Okay, in the last round of comments I realized that using update() to ensure consistency was better, but if you want, I'll replace it with fbo = requiresFbo() wherever required.

That being said, it cannot be applied in most nodes. The reason being, they simply use the FBOs without creating them. Point in case, the DeferredPointLightsNode. It does not use a requiresFbo(), so we have to use update() or manually do it. I decided to adopt the cheeky way ;)

File: engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java (hunk 8 above)
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java]: The compiler did not throw a warning for redundant variable, so I just assumed that the variable was used in process().. Turns out it wasn't.

As it turns out, I converted this to a update() call, so that the readOnlyGBufferFbo.attachDepthBufferTo(refractiveReflectiveFbo); part can also run, even for the first frame. That, and I made the two FBOs local to the update() method, since we don't use them anywhere else.

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 10 above)
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: I'll add the comment as you suggest for now.

However, and I *might* be wrong here, I don't think it's particularly required, mainly because the enum is FboTextureTypes, and not FboAttachmentTypes, and depthStencilRboId is definitely not the former.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java (hunk 23 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java]: Sure. But the question is: why does it not use the requiresFbo() method? Because if a node binds an FBO, it surely requires it.

File: engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java (hunk 10 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/stateChanges/SetInputTextureFromFBO.java]: Giving a little bit more information, even if it's not strictly required, doesn't hurt.

File: engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java [hunk 25]
```diff
@@ -137,8 +135,8 @@ public void process() {
 
     @Override
     public void update() {
-        readOnlyGBufferFbo = displayResolutionDependentFBOs.get(READONLY_GBUFFER);
-        refractiveReflectiveFbo = displayResolutionDependentFBOs.get(REFRACTIVE_REFLECTIVE);
+        FBO readOnlyGBufferFbo = displayResolutionDependentFBOs.get(READONLY_GBUFFER);
+        FBO refractiveReflectiveFbo = displayResolutionDependentFBOs.get(REFRACTIVE_REFLECTIVE);
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/RefractiveReflectiveBlocksNode.java]: Why did you do this?

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java (hunk 23 above)
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java]: That's an interesting question, and I think the answer is pretty simple - since it's a static renderGraph right now (if you can call it a graph), and so the programmer just used the requiresFbo() in the first Node that uses the particular Fbo. That, and all nodes assume the default Fbos don't have to explicitly asked-for as requirement.

I inherited the problem :P

**Comment by Cervator** [comment]
Tested again and merged! Huzzah!

**Comment by emanuele3d** [comment]
Well done @vampcat!
